src/ucp/proto/proto_multi.c

```diff
@@ -151,6 +151,91 @@ ucp_proto_multi_init_flush_sys_dev_mask(const ucp_rkey_config_key_t *key)
     return UCS_BIT(key->sys_dev & ~UCP_SYS_DEVICE_FLUSH_BIT);
 }
 
+static const uct_tl_resource_desc_t *
+ucp_proto_multi_get_tl_rsc(const ucp_proto_init_params_t *params,
+                           ucp_lane_index_t lane)
+{
+    ucp_rsc_index_t rsc_index = ucp_proto_common_get_rsc_index(params, lane);
+    return &params->worker->context->tl_rscs[rsc_index].tl_rsc;
+}
+
+static uct_device_type_t
+ucp_proto_multi_get_dev_type(const ucp_proto_init_params_t *params,
+                             ucp_lane_index_t lane)
+{
+    return ucp_proto_multi_get_tl_rsc(params, lane)->dev_type;
+}
+
+static const char *
+ucp_proto_multi_get_dev_name(const ucp_proto_init_params_t *params,
+                             ucp_lane_index_t lane)
+{
+    return ucp_proto_multi_get_tl_rsc(params, lane)->dev_name;
+}
+
+static ucp_lane_index_t ucp_proto_multi_filter_net_devices(
+        ucp_lane_index_t num_lanes, const ucp_proto_init_params_t *params,
+        const ucp_proto_common_tl_perf_t *lanes_perf, int fixed_first_lane,
+        ucp_lane_index_t *lanes, ucp_proto_perf_node_t **lanes_perf_nodes)
+{
+    double max_bandwidth;
+    ucp_lane_index_t i, lane, num_identical_devs, seed, num_filtered_lanes;
+    ucp_lane_map_t lane_map;
+    const char *dev_name;
+    const char *dev_names[UCP_PROTO_MAX_LANES];
+    const uct_tl_resource_desc_t *tl_rsc;
+
+    for (lane_map = 0, max_bandwidth = 0, i = 0; i < num_lanes; ++i) {
+        lane = lanes[i];
+        if (ucp_proto_multi_get_dev_type(params, lane) != UCT_DEVICE_TYPE_NET) {
+            continue;
+        }
+
+        lane_map     |= UCS_BIT(lane);
+        max_bandwidth = ucs_max(max_bandwidth, lanes_perf[lane].bandwidth);
+    }
+
+    num_identical_devs = 0;
+    ucs_for_each_bit(lane, lane_map) {
+        if (fabsf(lanes_perf[lane].bandwidth - max_bandwidth) >
+            UCP_PROTO_PERF_EPSILON) {
+            continue;
+        }
+
+        dev_name = ucp_proto_multi_get_dev_name(params, lane);
```

Comment:
maybe better use sys_devs of each lane instead of dev_name

Reply:
It might be enough. However, I thought it would be more reliable this way, since theoretically two transports can correspond to one physical device. E.g: `tcp/ib0`, and `rc_mlx5\`mlx5_0:1`.

Reply:
So, this confirms my previous comment. Two devices has the same sys_dev=1, different names.

Reply:
Fixed.